### PR TITLE
interface: refactored chat window into header, window, and scroll container

### PR DESCRIPTION
**Pull request by tacryt-socryp**
Maintained all functionality, just split it into different files to organize things.

As far as I can tell from testing, everything works in an equivalent manner. Big change, so please help out with testing!

### CODE REVIEW

File: pkg/interface/src/apps/chat/components/lib/chat-messages.tsx [hunk 1]
```diff
@@ -0,0 +1,96 @@
+import React, { Component, Fragment } from "react";
+import moment from "moment";
+
+import { Message } from "./message";
+
+type IMessage = Envelope & { pending?: boolean };
+
+
+export class ChatMessages extends Component {
```
Comment (liam-fitzgerald) [inline pkg/interface/src/apps/chat/components/lib/chat-messages.tsx:9]: typing props would be nice. e.g. `extends Component<ChatMessagesProps>`

File: pkg/interface/src/apps/chat/components/lib/chat-header.js [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+              {title}
+            </h2>
+          </Link>
+          <ChatTabBar
+            {...props}
+            station={props.station}
+            numPeers={group.length}
+            isOwner={deSig(props.match.params.ship) === window.ship}
+            popout={props.popout}
+            api={props.api}
```
Comment (liam-fitzgerald) [inline pkg/interface/src/apps/chat/components/lib/chat-header.js:62]: some of these props are made redundant by `{..props}`

File: pkg/interface/src/apps/chat/components/lib/chat-messages.tsx [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+      if (unread > 0 && i === unread - 1) {
+        return (
+          <Fragment key={msg.uid}>
+            {messageElem}
+            <div
+              ref={this.setUnreadMarker}
+              className="mv2 green2 flex items-center f9"
+            >
+              <hr className="dn-s ma0 w2 b--green2 bt-0" />
+              <p className="mh4">New messages below</p>
```
Comment (liam-fitzgerald) [inline pkg/interface/src/apps/chat/components/lib/chat-messages.tsx:65]: might be nice to refactor into `renderUnreadDivider()` and `renderDayBreakDivider()` or the like.

File: pkg/interface/src/apps/chat/components/lib/chat-header.js [hunk 4]
```diff
@@ -0,0 +1,68 @@
+import React, { Component, Fragment } from "react";
+import { Link } from "react-router-dom";
+
+import { ChatTabBar } from "./chat-tabbar";
+import { SidebarSwitcher } from "../../../../components/SidebarSwitch";
+import { deSig } from "../../../../lib/util";
+
+
+export class ChatHeader extends Component {
```
Comment (liam-fitzgerald) [inline pkg/interface/src/apps/chat/components/lib/chat-header.js:9]: TS would be nice, but not a dealbreaker

File: pkg/interface/src/apps/chat/components/lib/chat-header.js [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+    const { props } = this;
+    const isinPopout = props.popout ? "popout/" : "";
+    const group = Array.from(props.group.members);
+    let title = props.station.substr(1);
+    if (props.association && "metadata" in props.association) {
+      title =
+        props.association.metadata.title !== ""
+          ? props.association.metadata.title
+          : props.station.substr(1);
+    }
```
Comment (liam-fitzgerald) [inline pkg/interface/src/apps/chat/components/lib/chat-header.js:21]:
```suggestion
    if (props.association && 
      "metadata" in props.association && 
      props.association.metadata.tile !== "") {
      title = props.association.metadata.title
    }
```

**Comment by liam-fitzgerald** [comment]
> * need to set scroll lock properly when inactive (do we actually need this feature? seems unnecessarily bespoke 


cc: @urcades 
FWIW, I really like the current scrolling behaviour and I'd be hesistant to change it, given the number of iterations we did on this.

File: pkg/interface/src/apps/chat/components/lib/chat-header.js (hunk 2 above)
Comment (tacryt-socryp) [reply pkg/interface/src/apps/chat/components/lib/chat-header.js:62]: I don't want to pass in unnecessary props.

**Comment by urcades** [comment]
In order to make a semblance of a call here, I need to know what's precisely meant by:

> need to set scroll lock properly when inactive (do we actually need this feature? seems unnecessarily bespoke

Specifically, what a "scroll lock" is — and what we meant when we say "current scrolling behaviour".

Is this discussion with respect to the **literal mechanic** how we 'freeze' people's chat where they last left off, and keep them in-place for when they revisit a chat? 
Is this discussion with respect to the "_n_ new messages since last visit" bar that populates a chat when this mechanism is triggered? Are we discussing the whole thing here?

If we're discussing the mechanic as a broad structure: I'd argue for its inclusion in this refactor seeing as "respecting the scroll-journey of users" is a frequently abused anti-pattern various feed-based interfaces leverage, generally for the purpose of showing ads and algorithmic reshuffled feeds and the like. Seeing as we're not about that, I feel rather strongly this is a mechanic we should try to keep intact — not only in chat, but as a pattern replicated in a variety of scrolling contexts we might end up building in modules.

If this isn't what we're discussing here, please let me know!

**Comment by urcades** [comment]
Deeply into this PR regardless, these separations seem incredibly good/sane for the purposes of further development — we should replicate this in all other modules we develop.

**Comment by matildepark** [comment]
> need to set scroll lock properly when inactive (do we actually need this feature? seems unnecessarily bespoke

I also don't follow this. Currently our scroll doesn't lock properly (see #2874) and slams to the bottom regardless of where you are. 

We basically want the scroll to float without being pinned to bottom unless we're at the bottom or if we send a message. If there's logic in there that seems very bespoke, it also likely is currently non-functional, I think.

File: pkg/interface/src/apps/chat/components/lib/chat-header.js (hunk 2 above)
Comment (liam-fitzgerald) [reply pkg/interface/src/apps/chat/components/lib/chat-header.js:62]: you're passing them in anyway with `{...props}` on line 57.

Comment (tacryt-socryp) [reply pkg/interface/src/apps/chat/components/lib/chat-header.js:62]: Good point, got rid of them!

File: pkg/interface/src/apps/chat/components/lib/chat-messages.tsx (hunk 1 above)
Comment (tacryt-socryp) [reply pkg/interface/src/apps/chat/components/lib/chat-messages.tsx]: Killed this whole file.

File: pkg/interface/src/apps/chat/components/lib/chat-header.js (hunk 4 above)
Comment (tacryt-socryp) [reply pkg/interface/src/apps/chat/components/lib/chat-header.js:9]: TS seems excessive for a component with no logic.

**Comment by matildepark** [comment]
Some very weird scroll behaviour. Scroll up and it immediately pulls you back down a page at a time.
Aug-06-2020 16-02-41

Brave.

Also the "mark as read" indicator doesn't dismiss once you hit bottom; only once you click it manually.

**Comment by tacryt-socryp** [comment]
I've only tested in Safari. Will look at Brave behavior ☹️

**Comment by matildepark** [comment]
Well it would probably affect all Blink browsers. I bet Firefox works great.

**Comment by tacryt-socryp** [comment]
@matildepark fixed!

**Comment by matildepark** [comment]
Say I want to see what happened yesterday; I scroll a step into the backlog and look at the mid-point of it. I get forced out and bumped back down into the latest page. If I scroll *really* far this doesn't happen but otherwise this is pretty consistent.

Aug-06-2020 17-26-11

**Comment by tacryt-socryp** [comment]
@matildepark Fixed it, I had introduced a bug with resetting pageNum to 1 in my last commit.

**Comment by matildepark** [comment]
Works on Brave. Scroll load in Firefox works, with the caveat that it jumps *straight* to the top of the loaded backlog page. Don't know if scroll slamming is averted in Firefox. But mostly addresses #2864, maybe put "Fixes ####" in one of these commits so the issue auto-closes once it hits master.

**Comment by tacryt-socryp** [comment]
I'll just manually close it myself..

**Pull request merged**